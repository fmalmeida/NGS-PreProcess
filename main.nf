#!/usr/bin/env nextflow
nextflow.enable.dsl=2

/*
 * Generic Pipeline for preprocessing ngs data
 */

/*
 * Include functions
 */
include { helpMessage } from './nf_functions/help.nf'
include { exampleMessage } from './nf_functions/examples.nf'
//include { paramsCheck } from './nf_functions/paramsCheck.nf'
include { configMessage; illuminaMessage; ontMessage; pacbioMessage; logMessage } from './nf_functions/logMessages.nf'

/*
 * Check parameters
 */
//paramsCheck()
params.help = false
if (params.help){
  helpMessage()
  exit 0
}
params.examples = false
if (params.examples){
  exampleMessage()
  exit 0
}

/*
 * Download configuration files if requested
 */
params.get_full_config = false
params.get_illumina_config = false
params.get_ont_config = false
params.get_pacbio_config = false

if (params.get_full_config) {
  new File("ngs-preprocess.config").write(new URL ("https://github.com/fmalmeida/ngs-preprocess/raw/master/nextflow.config").getText())
  configMessage()
  exit 0
}

if (params.get_illumina_config) {
  new File("illumina_data.config").write(new URL ("https://github.com/fmalmeida/ngs-preprocess/raw/master/configuration_example/illumina_data.config").getText())
  illuminaMessage()
  exit 0
}

if (params.get_ont_config) {
  new File("ont_data.config").write(new URL ("https://github.com/fmalmeida/ngs-preprocess/raw/master/configuration_example/ont_data.config").getText())
  ontMessage()
  exit 0
}

if (params.get_pacbio_config) {
  new File("pacbio_data.config").write(new URL ("https://github.com/fmalmeida/ngs-preprocess/raw/master/configuration_example/pacbio_data.config").getText())
  pacbioMessage()
  exit 0
}

/*
 * Load general parameters and establish defaults
 */
params.outdir = 'output'
params.threads = 2

/*
 * Parameters for short reads
 */
params.shortreads = ''
params.shortreads_type = 'paired' //paired or single
params.clip_r1 = 0
params.clip_r2 = 0
params.three_prime_clip_r1 = 0
params.three_prime_clip_r2 = 0
params.quality_trim = 20
params.lighter_execute = false
params.lighter_kmer = 21
params.lighter_genomeSize = false
params.lighter_alpha = false
params.flash_execute = false

/*
 * Parameters for longreads filtering
 */
params.lreads_min_quality = false
params.lreads_min_length  = false

/*
 * Parameters for nanopore longreads
 */
params.nanopore_fastq = ''
params.nanopore_is_barcoded = false
params.nanopore_sequencing_summary = ''

/*
 * Parameters for pacbio longreads
 */
params.pacbio_bamPath = ''
params.pacbio_h5Path = ''
params.pacbio_barcodes = ''
params.pacbio_barcode_design = 'same'
params.pacbio_get_hifi = false

/*
 * Define log message
 */
logMessage()

/*
 * Include modules
 */
include { porechop } from './modules/porechop.nf' params(outdir: params.outdir)

include { nanopack; nanopack as nanopack_hifi } from './modules/nanopack.nf' params(outdir: params.outdir)

include { lreads_filter; lreads_filter as lreads_filter_hifi } from './modules/lreads_filter.nf' params(outdir: params.outdir,
  lreads_min_length: params.lreads_min_length, lreads_min_quality: params.lreads_min_quality)

include { pycoQC } from './modules/pycoQC.nf' params(outdir: params.outdir)

include { pacbio_bam2fastq } from './modules/pacbio_bam2fastq.nf' params(outdir: params.outdir,
  pacbio_barcodes: params.pacbio_barcodes, pacbio_barcode_design: params.pacbio_barcode_design,
  threads: params.threads)

include { pacbio_bam2hifi } from './modules/pacbio_bam2hifi.nf' params(outdir: params.outdir,
  pacbio_barcodes: params.pacbio_barcodes, pacbio_barcode_design: params.pacbio_barcode_design,
  threads: params.threads)

include { pacbio_h52bam } from './modules/pacbio_h52bam.nf' params(outdir: params.outdir)

include { fastqc } from './modules/fastqc.nf' params(outdir: params.outdir,
  shortreads_type: params.shortreads_type)

include { trimgalore } from './modules/trimgalore.nf' params(outdir: params.outdir,
  shortreads_type: params.shortreads_type, clip_r1: params.clip_r1,
  clip_r2: params.clip_r2, three_prime_clip_r1: params.three_prime_clip_r1,
  three_prime_clip_r2: params.three_prime_clip_r2, quality_trim: params.quality_trim)

include { lighter } from './modules/lighter.nf' params(outdir: params.outdir,
  lighter_kmer: params.lighter_kmer, lighter_alpha: params.lighter_alpha,
  shortreads_type: params.shortreads_type, lighter_genomeSize: params.lighter_genomeSize)

include { flash } from './modules/flash.nf' params(outdir: params.outdir)

/*
 * Define custom workflows
 */
workflow nanopore_nf {
  take:
    reads
    threads
    barcode
  main:
    porechop(reads, threads, barcode)
    nanopack(porechop.out[0].flatten(), threads)
    if (params.lreads_min_length || params.lreads_min_quality) {
      lreads_filter(porechop.out[0].flatten())
    }
}

workflow pycoQC_nf {
  take:
    input
  main:
    pycoQC(input)
}

workflow pacbio_bam_nf {
  take:
    subreads
    barcodes
    threads
  main:
    pacbio_bam2fastq(subreads, barcodes)
    nanopack(pacbio_bam2fastq.out[0].flatten(), threads)
    if (params.lreads_min_length || params.lreads_min_quality) {
      lreads_filter(pacbio_bam2fastq.out[0].flatten())
    }

    // User wants to get hifi?
    if (params.pacbio_get_hifi) {
      pacbio_bam2hifi(subreads, barcodes)
      nanopack_hifi(pacbio_bam2hifi.out[0].flatten(), threads)
      if (params.lreads_min_length || params.lreads_min_quality) {
        lreads_filter_hifi(pacbio_bam2hifi.out[0].flatten())
      }
    }
}

workflow pacbio_bas_nf {
  take:
    h5bas
    h5bas_dir
    barcodes
    threads
  main:
    pacbio_h52bam(h5bas, h5bas_dir)
    bams = pacbio_h52bam.out[0]
    pacbio_bam2fastq(bams, barcodes)
    nanopack(pacbio_bam2fastq.out[0].flatten(), threads)
    if (params.lreads_min_length || params.lreads_min_quality) {
      lreads_filter(pacbio_bam2fastq.out[0].flatten())
    }

    // User wants to get hifi?
    if (params.pacbio_get_hifi) {
      pacbio_bam2hifi(bams, barcodes)
      nanopack_hifi(pacbio_bam2hifi.out[0].flatten(), threads)
      if (params.lreads_min_length || params.lreads_min_quality) {
        lreads_filter_hifi(pacbio_bam2hifi.out[0].flatten())
      }
    }
}

workflow shortreads_nf {
  take:
    preads
    sreads
    threads
  main:
    fastqc(preads, sreads, threads)
    trimgalore(preads, sreads, threads)

    // Paired
    if (params.shortreads_type == 'paired') {
      if (params.lighter_execute && params.flash_execute) {
        lighter(trimgalore.out[0], threads)
        flash(lighter.out[0], threads)
      } else if (params.lighter_execute && !params.flash_execute) {
        lighter(trimgalore.out[0], threads)
      } else if (!params.lighter_execute && params.flash_execute) {
        flash(trimgalore.out[0], threads)
      }
    }

    // Single
    if (params.shortreads_type == 'single') {
      if (params.lighter_execute) {
        lighter(trimgalore.out[1].flatten(), threads)
      }
    }
}

/*
 * Define main workflow
 */
workflow {
  /*
   * User has nanopore longreads
   */
  if (params.nanopore_fastq) {
    nanopore_nf(Channel.fromPath(params.nanopore_fastq), params.threads, params.nanopore_is_barcoded)
  }

  /*
   * User wants to render a report with pycoQC
   */
  if (params.nanopore_sequencing_summary) {
    pycoQC_nf(Channel.fromPath(params.nanopore_sequencing_summary))
  }

  /*
   * User has pacbio subreads in bam format
   */
  if (params.pacbio_bamPath) {
    pacbio_bam_nf(Channel.fromPath(params.pacbio_bamPath),
                  (params.pacbio_barcodes) ? Channel.fromPath(params.pacbio_barcodes) : Channel.value(''),
                  params.threads)
  }

  /*
   * User has pacbio subreads in legacy h5 (bas and bax) files
   */
  if (params.pacbio_h5Path) {
    pacbio_bas_nf(Channel.fromPath(params.pacbio_h5Path),
                  Channel.fromPath(params.pacbio_h5Path, type: 'dir'),
                  (params.pacbio_barcodes) ? Channel.fromPath(params.pacbio_barcodes) : Channel.value(''),
                  params.threads)
  }

  /*
   * User has short paired end reads
   */
  if (params.shortreads && params.shortreads_type == 'paired') {
    shortreads_nf(Channel.fromFilePairs(params.shortreads, flat: true, size: 2),
                  Channel.value(''), params.threads)
  }

  /*
   * User has short single end reads
   */
  if (params.shortreads && params.shortreads_type == 'single') {
    shortreads_nf(Channel.value(['', '', '']), Channel.fromPath(params.shortreads), params.threads)
  }
}

workflow.onComplete {
    println ""
    println "Pipeline completed at: $workflow.complete"
    println "Execution status: ${ workflow.success ? 'OK' : 'failed' }"
    println "Execution duration: $workflow.duration"
    println "Thank you for using fmalmeida/ngs-preprocess pipeline!"
}
