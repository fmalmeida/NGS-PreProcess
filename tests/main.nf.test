nextflow_pipeline {

    name "Test Workflow main.nf"
    script "main.nf"

    test("Testing workflow - Should run without fails") {

        when {
            params {
                sra_ids    = 'https://github.com/fmalmeida/test_datasets/raw/main/sra_ids.txt'
                output     = "${outputDir}/test_output"
                max_cpus   = 2
                max_memory = '6.GB'
                max_time   = '6.h'
            }
        }

        then {
            
            assertAll(
                // All good?
                {assert workflow.success},

                // All there?
                {assert path("${outputDir}/test_output").list().size() == 5},
                {assert path("${outputDir}/test_output/preprocessing_outputs").list().size() == 3},

                // Check if files are the same
                {assert snapshot(
                    workflow,
                    path( "${outputDir}/test_output/final_output" )
                ).match()}
            )
        
        }

    }

}
