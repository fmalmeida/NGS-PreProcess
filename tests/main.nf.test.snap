{
    "Testing workflow - Should run without fails": {
        "content": [
            {
                "stderr": [
                    
                ],
                "errorReport": "",
                "exitStatus": 0,
                "failed": false,
                "stdout": [
                    "",
                    "Pipeline completed at: 2024-05-06T07:56:14.928664372Z",
                    "Execution status: OK",
                    "Execution duration: 9m 50s",
                    "Thank you for using fmalmeida/ngs-preprocess pipeline!"
                ],
                "errorMessage": "",
                "trace": {
                    "tasksFailed": 0,
                    "tasksCount": 12,
                    "tasksSucceeded": 12
                },
                "name": "workflow",
                "success": true
            },
            [
                [
                    "SRR19901041.filtered.fq.gz:md5,1d7a8e18db3b1f7ac0bfc1f5cf1daddf"
                ],
                [
                    "ERR7339227.filtered.fq.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                [
                    "ERR9850226_R1.preprocessed.fq.gz:md5,69f8bd550fbf99d81f4a110734f4803a",
                    "ERR9850226_R2.preprocessed.fq.gz:md5,7543436f1614ae814337a9b123d26376"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-06T07:56:15.500768292"
    }
}